consensus-types/blocks/execution.go

```diff
@@ -980,23 +945,14 @@ func IsEmptyExecutionData(data interfaces.ExecutionData) (bool, error) {
 		return false, nil
 	}
 
-	drs, err := data.DepositReceipts()
-	switch {
-	case errors.Is(err, consensus_types.ErrUnsupportedField):
-	case err != nil:
-		return false, err
-	default:
+	epe, postElectra := data.(interfaces.ExecutionDataElectra)
+	if postElectra {
+		drs := epe.DepositReceipts()
```

Comment:
This pattern assumes that every `interfaces.ExecutionData` in future forks, that also satisfies `ExecutionDataElectra` will have to be handled in the same way. Say for example that `interfaces.ExecutionDataEpbs` satisfies the Electra interface, it requires different handling of deposits but it requires the same handling of 7002 withdrawals, etc. If we explicitly change it to not satisfy `ExecutionDataElectra` and satisfy a different `ExectutionDataEpbs` then we would have to copy paste the withdrawal handling. If we keep it satisfying `ExecutionDataElectra` then we would have to deal with deposits with an if-else `data.(interfaces.ExecutionDataEpbs)` within this block right?

Reply:
Copy/paste from team chat:

If the ssz is the same, then it's fine for `ExecutionDataEpbs` to satisfy `ExecutionDataElectra`. Satisfying the interface only dictates what kind of value are available, it does not dictate how the data should be processed. That should always be based on the slot (or version, which is based on the slot).

SSZ changes in nested types is similar to what we just went through with `Attestations`. In that case our solution was to move attestations to an interface. A benefit of this kind of composition is that we would only need to implement the logic to deal with the interface in 2 places.If we do want to actually deprecate (remove) a field, then yes one option would be to break the chain of interface composition. Another would be to backport an error to that fields signature. Another would be to allow the getter to return a nil value and rely on the caller to be fork-aware. Another option is to rely less on these big interfaces and move processing functions to accept narrower interfaces. For example, epbs could take:
```
type EpbsWithdrawalGetter interface {
    GetWithdrawals() *eth.PBSWithdrawal
}
```
whereas other methods would take:
```
type WithdrawalsGetter interface {
    GetWithdrawals() []byte
}
```
and then perform assertions on those instead of the whole big type. A problem in both cases is that now if we want block to have something like an `ExecutionData` type that the container could use to abstractly carry the payload. This type would have to become /increasingly/ abstract (some kind of `ConsensusType` interface that would have ssz methods etc, which is asserted downstream to narrower interfaces). Otherwise we would also break the chain of composition in the container (block body).

*TL;DR* this approach handles all the previous fork changes pretty well, with attestations being the only notable exception. If we see more extreme changes across forks I think the best way to deal with that is probably more abstract interfaces for nested types, and narrow interfaces used by consumers of those fields. We'll need to keep iterating on it to see what feels most maintainable in practice.